src/Bicep.Core/Analyzers/Linter/Rules/OutputsShouldNotContainSecretsRule.cs

```diff
@@ -88,23 +91,108 @@ public OutputValueVisitor(OutputsShouldNotContainSecretsRule parent, List<IDiagn
+                }
+
+                accessExpressionStack.Push(syntax);
+                base.VisitPropertyAccessSyntax(syntax);
+                accessExpressionStack.Pop();
+            }
+
+            public override void VisitArrayAccessSyntax(ArrayAccessSyntax syntax)
+            {
+                foreach (var pathToSecureComponent in FindPathsToSecureComponents(model.GetTypeInfo(syntax)))
+                {
+                    string foundMessage = string.Format(CoreResources.OutputsShouldNotContainSecretsSecureParam, syntax.ToTextPreserveFormatting() + pathToSecureComponent);
+                    this.diagnostics.Add(parent.CreateDiagnosticForSpan(diagnosticLevel, syntax.IndexExpression.Span, foundMessage));
+                }
+
+                accessExpressionStack.Push(syntax);
+                base.VisitArrayAccessSyntax(syntax);
+                accessExpressionStack.Pop();
+            }
+
+            private IEnumerable<string> FindPathsToSecureComponents(TypeSymbol type)
+                => FindPathsToSecureComponents(type, "", ImmutableHashSet<TypeSymbol>.Empty);
+
+            private IEnumerable<string> FindPathsToSecureComponents(TypeSymbol type, string path, ImmutableHashSet<TypeSymbol> visited)
+            {
+                // types can be recursive. cut out early if we've already seen this type
+                if (visited.Contains(type))
                 {
-                    if (param.IsSecure())
+                    yield break;
+                }
+
+                if (type.ValidationFlags.HasFlag(TypeSymbolValidationFlags.IsSecure))
+                {
+                    yield return path;
+                    // if we encounter a type that has been explicitly flagged as secure, stop visiting its components
+                    yield break;
+                }
+
+                if (type is UnionType union)
+                {
+                    foreach (var variantPath in union.Members.SelectMany(m => FindPathsToSecureComponents(m.Type, path, visited.Add(type))))
                     {
-                        string foundMessage = string.Format(CoreResources.OutputsShouldNotContainSecretsSecureParam, syntax.Name.IdentifierName);
-                        this.diagnostics.Add(parent.CreateDiagnosticForSpan(diagnosticLevel, syntax.Name.Span, foundMessage));
+                        yield return variantPath;
                     }
                 }
 
-                base.VisitVariableAccessSyntax(syntax);
+                // if the expression being visited is dereferencing a specific property or index of this type, we shouldn't warn if the type under inspection
+                // *contains* properties or indices that are flagged as secure. We will have already warned if those have been accessed in the expression, and
+                // if they haven't, then the value dereferenced isn't sensitive
+                if (accessExpressionStack.Count == 0)
```

Comment:
I'm probably wrong but want to check: wouldn't the `accessExpressionStack.Count` always be greater than 0 since we only call this method after pushing to the stack? Confused how the below switch statements would be reached

Reply:
Adding some of the examples above to the code would help a lot, thanks!